### PR TITLE
convert data api spec to raml

### PR DESCRIPTION
**Pull request by tardyp**
raml is a metalanguage to describe REST apis. http://raml.org/

The claim is to be the successor from swagger and blueprint, with focus being
on a easy to read source code.

My goal is is to evntually have all the rtype and most of data api doc
written in raml, with a bunch of scripts to convert the raml into

- sphinx doc for resource types and data api (regenerate the same doc as we currently have)
- sphinx doc for rest api (we don't have that right now, and ask our users to manually derive it from data api)
- sphinx doc for mq (which currently is mostly outdated, as it is derived from the rtype api http://trac.buildbot.net/ticket/3214)

All the raml file in this commit are purely derived from the typing metadata associated with endpoints declaration in the data api.

The goal is to enhance the raml files so that they include the rst doc we currently have in rtype-*.rst
and to have a unit test to automatically verify coherency b/w raml doc, and data/*.py

I dont think I want to rewrite the data validation stuff directly using raml, this can be an eventual goal, but not really needed at the moment

Another usage of this is also to generate metadata for the data-module, which currently hardcodes some stuff in the coffee.

### CODE REVIEW

**Comment by tardyp** [comment]
@rutsky this is something that can clash a little bit with what you are doing.

Eventually, this work will generate docs/developer/rtype-*

This is however very little overhead, as this is really only doc, which will be translated easily via search/replace

**Comment by jaredgrubb** [comment]
This is really cool. I havent seen raml before.

**Comment by rutsky** [comment]
@tardyp I see. I haven't touched Data API and it's docs yet, I will do them after your changes will be merged (that is planned on next week, according to the last meeting resolution).

**Comment by tardyp** [comment]
continuing this work...

see work in progress result:
https://circle-artifacts.com/gh/buildbot/buildbot/252/artifacts/0/home/ubuntu/buildbot/master/docs/_build/html/developer/rest.html

**Comment by tardyp** [comment]
I finished writing the raml files (actually copy pasting from the current doc to the new raml description)

I would apreciate review, comments on the raml files.

current doc generation results is here:
https://circle-artifacts.com/gh/buildbot/buildbot/276/artifacts/0/home/ubuntu/buildbot/master/docs/_build/html/developer/rest.html

I'll work tomorrow on enhancing that including:
endpoint parameters doc
data types doc
jsonrpc2 "control" doc

**Comment by Stibbons** [comment]
+1

**Comment by tardyp** [comment]
new version, only remaining is json rpc doc.

https://circle-artifacts.com/gh/buildbot/buildbot/288/artifacts/0/home/ubuntu/buildbot/master/docs/_build/html/developer/rest.html

**Comment by sa2ajj** [comment]
```
exceptions.ImportError: No module named ramlfications
```

**Comment by sa2ajj** [comment]
I really would like to see this in

**Comment by tardyp** [comment]
@sa2ajj I will rebase this after the beta8 PR (it is nearly ready, apart from the ramlfications dependencies)

**Comment by sa2ajj** [comment]
yay :)

File: master/docs/developer/rest.rst [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+            {{raml.format_json(example, indent=4*2)}}
+
+    {% endfor %}
+    {% endif %}
+
+    {{type.description}}
+    {% endif %}
+    {% if name in raml.endpoints_by_type -%}
+    Endpoints
```
Comment (Stibbons) [inline master/docs/developer/rest.rst]: look like a endfor is missing here. the endfor line 236 is about the type (``{% for name, type in raml.types.items()|sort %}``)

Comment (Stibbons) [inline master/docs/developer/rest.rst]: forgot my comment, you print endpoints by types

**Pull request merged**